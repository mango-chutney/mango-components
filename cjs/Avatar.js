"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Avatar = Avatar;
exports.default = exports.createComponent = exports.createStyledComponents = exports.defaultStyleProps = void 0;

require("core-js/modules/es6.object.assign");

require("core-js/modules/web.dom.iterable");

require("core-js/modules/es6.array.iterator");

require("core-js/modules/es6.object.keys");

require("core-js/modules/es6.function.name");

var React = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _polished = require("polished");

var _constants = require("./constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var defaultStyleProps = {
  backgroundColor: '#00b4ff',
  fontWeight: _constants.fontWeights.semibold
};
exports.defaultStyleProps = defaultStyleProps;
var colorBackgrounds = ['#00d161', '#00b4ff', '#ff3add', '#ffca00', '#8151f3', '#ff7741'];

function selectColor(namespace) {
  var hash = 0;

  for (var i = 0; i < namespace.length; i += 1) {
    // eslint-disable-next-line no-bitwise
    hash = (hash << 5) - hash + namespace.charCodeAt(i); // eslint-disable-next-line no-bitwise

    hash |= 0; // Convert to 32bit integer
  }

  return colorBackgrounds[Math.abs(hash) % colorBackgrounds.length];
}

var createStyledComponents = function createStyledComponents(styleProps) {
  var AvatarComponent = _styledComponents.default.div.withConfig({
    componentId: "s1ykk4fe-0"
  })(["background-color:", ";background-position:center;background-repeat:no-repeat;background-size:cover;border-radius:50%;color:", ";font-size:", ";font-weight:", ";height:", ";line-height:", ";text-align:center;text-transform:uppercase;width:", ";"], styleProps.backgroundColor, _constants.palette.white, function (_ref) {
    var fontSize = _ref.fontSize;
    return fontSize;
  }, _constants.fontWeights.semibold, function (_ref2) {
    var width = _ref2.width;
    return width;
  }, function (_ref3) {
    var width = _ref3.width;
    return width;
  }, function (_ref4) {
    var width = _ref4.width;
    return width;
  });

  return {
    AvatarComponent: AvatarComponent
  };
};

exports.createStyledComponents = createStyledComponents;

function Avatar(props) {
  var AvatarComponent = props.AvatarComponent,
      backgroundColor = props.backgroundColor,
      backgroundImage = props.backgroundImage,
      children = props.children,
      name = props.name,
      style = props.style,
      width = props.width,
      rest = _objectWithoutProperties(props, ["AvatarComponent", "backgroundColor", "backgroundImage", "children", "name", "style", "width"]);

  return React.createElement(AvatarComponent, _extends({
    backgroundColor: backgroundColor,
    style: _objectSpread({
      backgroundColor: name && selectColor(name),
      backgroundImage: backgroundImage && "url(".concat(backgroundImage, ")")
    }, style),
    width: width
  }, rest), !backgroundImage && name && name.charAt(0) || children);
}

Avatar.defaultProps = {
  width: (0, _polished.rem)(46),
  fontSize: (0, _polished.rem)(18)
};

var createComponent = function createComponent() {
  var defaultStyledComponents = createStyledComponents(defaultStyleProps);
  return function (props) {
    return React.createElement(Avatar, _objectSpread({}, defaultStyledComponents, props));
  };
};

exports.createComponent = createComponent;
var _default = {
  defaultStyleProps: defaultStyleProps,
  createStyledComponents: createStyledComponents,
  createComponent: createComponent
};
exports.default = _default;