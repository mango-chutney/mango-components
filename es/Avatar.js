import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.array.iterator";
import "core-js/modules/es6.object.keys";
import "core-js/modules/es6.object.assign";
import "core-js/modules/es6.function.name";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import * as React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { palette, fontWeights } from './constants';
export var defaultStyleProps = {
  backgroundColor: '#00b4ff',
  fontWeight: fontWeights.semibold
};
var colorBackgrounds = ['#00d161', '#00b4ff', '#ff3add', '#ffca00', '#8151f3', '#ff7741'];

function selectColor(namespace) {
  var hash = 0;

  for (var i = 0; i < namespace.length; i += 1) {
    // eslint-disable-next-line no-bitwise
    hash = (hash << 5) - hash + namespace.charCodeAt(i); // eslint-disable-next-line no-bitwise

    hash |= 0; // Convert to 32bit integer
  }

  return colorBackgrounds[Math.abs(hash) % colorBackgrounds.length];
}

export var createStyledComponents = function createStyledComponents(styleProps) {
  var AvatarComponent = styled.div.withConfig({
    componentId: "s1ykk4fe-0"
  })(["background-color:", ";background-position:center;background-repeat:no-repeat;background-size:cover;border-radius:50%;color:", ";font-size:", ";font-weight:", ";height:", ";line-height:", ";text-align:center;text-transform:uppercase;width:", ";"], styleProps.backgroundColor, palette.white, function (_ref) {
    var fontSize = _ref.fontSize;
    return fontSize;
  }, fontWeights.semibold, function (_ref2) {
    var width = _ref2.width;
    return width;
  }, function (_ref3) {
    var width = _ref3.width;
    return width;
  }, function (_ref4) {
    var width = _ref4.width;
    return width;
  });
  return {
    AvatarComponent: AvatarComponent
  };
};
export function Avatar(props) {
  var AvatarComponent = props.AvatarComponent,
      backgroundColor = props.backgroundColor,
      backgroundImage = props.backgroundImage,
      children = props.children,
      name = props.name,
      style = props.style,
      width = props.width,
      rest = _objectWithoutPropertiesLoose(props, ["AvatarComponent", "backgroundColor", "backgroundImage", "children", "name", "style", "width"]);

  return React.createElement(AvatarComponent, _extends({
    backgroundColor: backgroundColor,
    style: Object.assign({
      backgroundColor: name && selectColor(name),
      backgroundImage: backgroundImage && "url(" + backgroundImage + ")"
    }, style),
    width: width
  }, rest), !backgroundImage && name && name.charAt(0) || children);
}
Avatar.defaultProps = {
  width: rem(46),
  fontSize: rem(18)
};
export var createComponent = function createComponent() {
  var defaultStyledComponents = createStyledComponents(defaultStyleProps);
  return function (props) {
    return React.createElement(Avatar, Object.assign({}, defaultStyledComponents, props));
  };
};
export default {
  defaultStyleProps: defaultStyleProps,
  createStyledComponents: createStyledComponents,
  createComponent: createComponent
};