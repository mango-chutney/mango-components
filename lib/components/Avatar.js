'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createComponent = exports.createStyledComponents = exports.defaultStyleProps = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.Avatar = Avatar;

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _polished = require('polished');

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var defaultStyleProps = exports.defaultStyleProps = {
  backgroundColor: '#00b4ff',
  fontWeight: _constants.fontWeights.semibold
};

var colorBackgrounds = ['#00d161', '#00b4ff', '#ff3add', '#ffca00', '#8151f3', '#ff7741'];

function selectColor(namespace) {
  var hash = 0;

  for (var i = 0; i < namespace.length; i += 1) {
    // eslint-disable-next-line no-bitwise
    hash = (hash << 5) - hash + namespace.charCodeAt(i);
    // eslint-disable-next-line no-bitwise
    hash |= 0; // Convert to 32bit integer
  }

  return colorBackgrounds[Math.abs(hash) % colorBackgrounds.length];
}

var createStyledComponents = exports.createStyledComponents = function createStyledComponents(styleProps) {
  var AvatarComponent = /*#__PURE__*/_styledComponents2.default.div.withConfig({
    componentId: 'xqbzju-0'
  })(['background-color:', ';background-position:center;background-repeat:no-repeat;background-size:cover;border-radius:50%;color:', ';font-size:', ';font-weight:', ';height:', ';line-height:', ';text-align:center;text-transform:uppercase;width:', ';'], styleProps.backgroundColor, _constants.palette.white, function (props) {
    return props.fontSize;
  }, _constants.fontWeights.semibold, function (props) {
    return props.width;
  }, function (props) {
    return props.width;
  }, function (props) {
    return props.width;
  });

  return { AvatarComponent: AvatarComponent };
};

function Avatar(props) {
  var AvatarComponent = props.AvatarComponent,
      backgroundColor = props.backgroundColor,
      backgroundImage = props.backgroundImage,
      children = props.children,
      name = props.name,
      style = props.style,
      width = props.width,
      rest = _objectWithoutProperties(props, ['AvatarComponent', 'backgroundColor', 'backgroundImage', 'children', 'name', 'style', 'width']);

  return React.createElement(
    AvatarComponent,
    _extends({
      backgroundColor: backgroundColor,
      style: _extends({
        backgroundColor: name && selectColor(name),
        backgroundImage: backgroundImage && 'url(' + backgroundImage + ')'
      }, style),
      width: width
    }, rest),
    !backgroundImage && name && name.charAt(0) || children
  );
}

Avatar.defaultProps = {
  width: (0, _polished.rem)(46),
  fontSize: (0, _polished.rem)(18)
};

var createComponent = exports.createComponent = function createComponent() {
  var defaultStyledComponents = createStyledComponents(defaultStyleProps);
  return function (props) {
    return React.createElement(Avatar, _extends({}, defaultStyledComponents, props));
  };
};

exports.default = {
  defaultStyleProps: defaultStyleProps,
  createStyledComponents: createStyledComponents,
  createComponent: createComponent
};