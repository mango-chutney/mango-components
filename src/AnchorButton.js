// @flow

import * as React from 'react';
import type { ReactComponentStyled as $ReactComponentStyled } from 'styled-components';
import type {
  $ComponentFactory,
  $MangoComponent,
  $StyledSubComponentsFactory,
} from './types';
import {
  createStyledComponents as createStyledButtonComponents,
  defaultStyleProps,
} from './Button';
import type { $StyledProps as $StyledButtonProps } from './Button';

export type $StyledProps = {
  ...$StyledButtonProps,
  href: string,
};

export type $Props = {
  ...$StyledProps,
  AnchorButtonComponent: React.ElementType,
};

export function AnchorButton(props: $Props) {
  const { children, href, AnchorButtonComponent, ...rest } = props;

  return (
    <AnchorButtonComponent href={href} {...rest}>
      {children}
    </AnchorButtonComponent>
  );
}

export { defaultStyleProps };

export const createStyledComponents: $StyledSubComponentsFactory<
  {
    AnchorButtonComponent: $ReactComponentStyled<*>,
  },
  typeof defaultStyleProps,
  // eslint-disable-next-line no-unused-vars
> = styleProps => {
  const AnchorButtonComponent = createStyledButtonComponents(
    defaultStyleProps,
  ).ButtonComponent.withComponent('a');

  return { AnchorButtonComponent };
};

export const createComponent: $ComponentFactory<$StyledProps> = () => {
  const defaultStyledComponents = createStyledComponents(defaultStyleProps);
  return (props: $StyledProps) => (
    <AnchorButton {...{ ...defaultStyledComponents, ...props }} />
  );
};

export default ({
  defaultStyleProps,
  createStyledComponents,
  createComponent,
}: $MangoComponent<typeof defaultStyleProps, $StyledProps>);
