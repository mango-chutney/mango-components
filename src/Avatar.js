// @flow

import * as React from 'react';
import styled from 'styled-components';
import type { ReactComponentStyled as $ReactComponentStyled } from 'styled-components';
import { rem } from 'polished';
import type {
  $ComponentFactory,
  $MangoComponent,
  $StyledSubComponentsFactory,
} from './types';
import { palette, fontWeights } from './constants';

export type $StyledProps = {
  backgroundColor?: string,
  backgroundImage?: string,
  children?: React.Node,
  fontSize?: string | number,
  name?: string,
  style?: {},
  width?: string | number,
};

export type $Props = {
  ...$StyledProps,
  AvatarComponent: React.ElementType,
};

export const defaultStyleProps: {|
  backgroundColor: ?string,
  fontWeight: ?string | ?number,
|} = {
  backgroundColor: '#00b4ff',
  fontWeight: fontWeights.semibold,
};

const colorBackgrounds = [
  '#00d161',
  '#00b4ff',
  '#ff3add',
  '#ffca00',
  '#8151f3',
  '#ff7741',
];

function selectColor(namespace) {
  let hash = 0;

  for (let i = 0; i < namespace.length; i += 1) {
    // eslint-disable-next-line no-bitwise
    hash = (hash << 5) - hash + namespace.charCodeAt(i);
    // eslint-disable-next-line no-bitwise
    hash |= 0; // Convert to 32bit integer
  }

  return colorBackgrounds[Math.abs(hash) % colorBackgrounds.length];
}

export const createStyledComponents: $StyledSubComponentsFactory<
  {
    AvatarComponent: $ReactComponentStyled<*>,
  },
  typeof defaultStyleProps,
> = styleProps => {
  const AvatarComponent = styled.div`
    background-color: ${styleProps.backgroundColor};
    background-position: center;
    background-repeat: no-repeat;
    background-size: cover;
    border-radius: 50%;
    color: ${palette.white};
    font-size: ${({ fontSize }) => fontSize};
    font-weight: ${fontWeights.semibold};
    height: ${({ width }) => width};
    line-height: ${({ width }) => width};
    text-align: center;
    text-transform: uppercase;
    width: ${({ width }) => width};
  `;

  return { AvatarComponent };
};

export function Avatar(props: $Props) {
  const {
    AvatarComponent,
    backgroundImage,
    children,
    name,
    style,
    ...rest
  } = props;

  return (
    <AvatarComponent
      style={{
        backgroundColor: name && selectColor(name),
        backgroundImage: backgroundImage && `url(${backgroundImage})`,
        ...style,
      }}
      {...rest}
    >
      {(!backgroundImage && name && name.charAt(0)) || children}
    </AvatarComponent>
  );
}

Avatar.defaultProps = {
  width: rem(46),
  fontSize: rem(18),
};

export const createComponent: $ComponentFactory<$StyledProps> = () => {
  const defaultStyledComponents = createStyledComponents(defaultStyleProps);
  return (props: $StyledProps) => (
    <Avatar {...{ ...defaultStyledComponents, ...props }} />
  );
};

export default ({
  defaultStyleProps,
  createStyledComponents,
  createComponent,
}: $MangoComponent<typeof defaultStyleProps, $StyledProps>);
