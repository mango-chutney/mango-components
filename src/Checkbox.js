// @flow

import * as React from 'react';
import styled, { css } from 'styled-components';
import type { ReactComponentStyled as $ReactComponentStyled } from 'styled-components';
import type { FieldProps as $FieldProps } from 'redux-form';
import { rem, darken } from 'polished';
import { palette, fontWeights } from './constants';
import type {
  $ComponentFactory,
  $MangoComponent,
  $StyledSubComponentsFactory,
} from './types';

export type $StyledProps = {
  id?: string,
  name: string,
  children: React.Node,
  label: string,
  checkboxColor: string,
  invalid?: boolean,
  disabled?: boolean,
} & $FieldProps;

export type $Props = {
  CheckboxBackgroundComponent: React.ComponentType<*>,
  CheckboxContainerComponent: React.ComponentType<*>,
  InputComponent: React.ComponentType<*>,
  LabelComponent: React.ComponentType<*>,
  WrapperComponent: React.ComponentType<*>,
} & $StyledProps;

export const defaultStyleProps: {|
  checkboxSize: number,
  checkboxColor: string,
|} = {
  checkboxSize: 20,
  checkboxColor: palette.primary,
};

export const createStyledComponents: $StyledSubComponentsFactory<
  {
    CheckboxBackgroundComponent: $ReactComponentStyled<*>,
    CheckboxContainerComponent: $ReactComponentStyled<*>,
    InputComponent: $ReactComponentStyled<*>,
    LabelComponent: $ReactComponentStyled<*>,
    WrapperComponent: $ReactComponentStyled<*>,
  },
  typeof defaultStyleProps,
> = styleProps => {
  const WrapperComponent = styled.div`
    margin: 1rem 0;
  `;

  const CheckboxContainerComponent = styled.div`
    position: relative;
    margin: 0;
    height: ${rem(styleProps.checkboxSize)};
    width: ${rem(styleProps.checkboxSize)};
    display: inline-block;
  `;

  // some label stuff in checkbox to get :checked property
  const CheckboxBackgroundComponent = styled.div`
    cursor: pointer;
    font-weight: ${fontWeights.semibold};
    display: inline-block;
    font-size: 1rem;

    ${({ invalid }) =>
      invalid &&
      css`
        color: ${palette.alert};
      `};

    min-height: ${rem(styleProps.checkboxSize)};
    line-height: ${rem(styleProps.checkboxSize)};
    position: relative;
    border: 1px solid transparent;
    display: block;

    &::before,
    &::after {
      width: ${rem(styleProps.checkboxSize)};
      height: ${rem(styleProps.checkboxSize)};
      cursor: pointer;
      content: '';
      display: inline-block;
      border-radius: 4px;
      vertical-align: middle;
    }

    &::before {
      border: 1px solid ${palette.border};
      background: ${palette.lightGray};
      margin-right: 1rem;

      ${({ invalid }) =>
        invalid &&
        css`
          border-color: ${palette.alert};
        `};

      ${({ disabled }) =>
        disabled &&
        css`
          background-color: ${darken(0.05, palette.lightGray)};
          cursor: not-allowed;
        `};
    }

    &::after {
      position: absolute;
      left: 0;
      border: 1px solid transparent;
      line-height: ${rem(styleProps.checkboxSize)};
      margin-top: ${rem(styleProps.checkboxSize * 0.1)};
      text-align: center;
      transform: scale(0);
    }
  `;

  const InputComponent = styled.input`
    width: 100%;
    height: 100%;
    position: absolute;
    padding: 0;
    margin: 0;
    top: 0;
    left: 0;
    z-index: 9001;
    opacity: 0;

    &[disabled] {
      cursor: not-allowed;
    }

    :checked + ${CheckboxBackgroundComponent} {
      &::before,
      &::after {
        top: 0;
        left: 0;
      }

      &::before {
        background: ${styleProps.checkboxColor};
        border-color: ${darken(0.05, styleProps.checkboxColor)};
        transition: background-color 0.3s;
      }

      &::after {
        content: '';
        background-image: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0naHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmcnIHhtbG5zOnhsaW5rPSdodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rJyBwcmVzZXJ2ZUFzcGVjdFJhdGlvPSd4TWlkWU1pZCcgd2lkdGg9JzE2LjUnIGhlaWdodD0nMTUnIHZpZXdCb3g9JzAgMCAxNi41IDE1Jz48cGF0aCBmaWxsPScjZmZmJyBmaWxsLXJ1bGU9J2V2ZW5vZGQnIGQ9J00xNi41MTUsMS4wNzEgQzE2LjM5OSwwLjgyNyAxNS45MTQsLTAuMDMwIDE1LjUxNSwwLjAwMSBDMTQuMjg5LDAuNzY1IDEyLjY5MywyLjM1NCAxMS42MzgsMy4zNjIgQzkuNTI4LDUuMzc4IDcuNTAzLDcuNDg2IDUuNTA4LDkuNjI0IEwxLjUxNiw1Ljg5NyBMMC4wMDUsNy40ODYgQzIuMjAwLDkuODM3IDQuNTY3LDEyLjM3MyA2LjUwNSwxNS4wMDAgQzguOTg2LDEwLjYzMiAxMy4wMzUsNC4zNzAgMTYuNTE1LDEuMDcxIEwxNi41MTUsMS4wNzEgWicvPjwvc3ZnPg==');
        background-position: 50% 40%;
        background-repeat: no-repeat;
        background-size: 70%;
        transition: all 0.25s;
        transform: scale(1);
      }
    }
  `;

  const LabelComponent = styled.label`
    margin-left: 1rem;
  `;

  return {
    CheckboxContainerComponent,
    CheckboxBackgroundComponent,
    LabelComponent,
    InputComponent,
    WrapperComponent,
  };
};

export function Checkbox({
  CheckboxBackgroundComponent,
  CheckboxContainerComponent,
  InputComponent,
  LabelComponent,
  WrapperComponent,
  children,
  id,
  input: { name, value, ...inputProps },
  label,
  meta,
  invalid,
  disabled,
  ...rest
}: $Props) {
  return (
    <WrapperComponent>
      <CheckboxContainerComponent>
        <InputComponent
          {...{
            ...rest,
            ...inputProps,
            id: id || name,
            name,
            checked: value,
            type: 'checkbox',
            invalid,
            disabled,
          }}
        />
        <CheckboxBackgroundComponent invalid={invalid} disabled={disabled} />
      </CheckboxContainerComponent>
      <LabelComponent htmlFor={id || name}>
        {children || label || ''}
      </LabelComponent>
    </WrapperComponent>
  );
}

export const createComponent: $ComponentFactory<$StyledProps> = () => {
  const defaultStyledComponents = createStyledComponents(defaultStyleProps);
  return (props: $StyledProps) => (
    <Checkbox {...{ ...defaultStyledComponents, ...props }} />
  );
};

export default ({
  defaultStyleProps,
  createStyledComponents,
  createComponent,
}: $MangoComponent<typeof defaultStyleProps, $StyledProps>);
