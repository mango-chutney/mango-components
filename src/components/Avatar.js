// @flow

import * as React from 'react';
import styled from 'styled-components';
import type { ReactComponentStyled } from 'styled-components';
import { rem } from 'polished';
import type {
  $ComponentFactory,
  $MangoComponent,
  $StyledSubComponentsFactory,
} from './types';
import { palette, fontWeights } from './constants';

export type $StyledProps = {
  children: React.Node,
  name?: string,
  backgroundColor?: string,
  backgroundImage?: string,
  size?: string | number,
  fontSize?: string | number,
};

export type $Props = {
  ...$StyledProps,
  AvatarComponent: React.ComponentType<*>,
};

export const defaultStyleProps: {|
  backgroundColor: ?string,
  fontWeight: ?string | ?number,
|} = {
  backgroundColor: '#00b4ff',
  fontWeight: fontWeights.semibold,
};

const colorBackgrounds = [
  '#00d161',
  '#00b4ff',
  '#ff3add',
  '#ffca00',
  '#8151f3',
  '#ff7741',
];

function selectColor(namespace) {
  let hash = 0;

  for (let i = 0; i < namespace.length; i += 1) {
    // eslint-disable-next-line no-bitwise
    hash = (hash << 5) - hash + namespace.charCodeAt(i);
    // eslint-disable-next-line no-bitwise
    hash |= 0; // Convert to 32bit integer
  }

  return colorBackgrounds[Math.abs(hash) % colorBackgrounds.length];
}

export const createStyledComponents: $StyledSubComponentsFactory<
  {
    AvatarComponent: ReactComponentStyled<*>,
  },
  typeof defaultStyleProps,
> = styleProps => {
  const AvatarComponent = styled.div`
    background-color: ${styleProps.backgroundColor};
    background-position: center;
    background-repeat: no-repeat;
    background-size: cover;
    border-radius: 50%;
    height: ${props => props.size};
    width: ${props => props.size};
    line-height: ${props => props.size};
    text-align: center;
    color: ${palette.white};
    font-weight: ${fontWeights.semibold};
    text-transform: uppercase;
    font-size: ${props => props.fontSize};
  `;

  return { AvatarComponent };
};

export function Avatar(props: $Props) {
  const {
    children,
    name,
    backgroundColor,
    backgroundImage,
    AvatarComponent,
    ...rest
  } = props;

  return (
    <AvatarComponent
      backgroundColor={backgroundColor}
      style={{
        backgroundColor: name && selectColor(name),
        backgroundImage: backgroundImage && `url(${backgroundImage})`,
      }}
      {...rest}
    >
      {(!backgroundImage && name && name.charAt(0)) || children}
    </AvatarComponent>
  );
}

Avatar.defaultProps = {
  size: rem(46),
  fontSize: rem(18),
};

export const createComponent: $ComponentFactory<$StyledProps> = () => {
  const defaultStyledComponents = createStyledComponents(defaultStyleProps);
  return (props: $StyledProps) => (
    <Avatar {...{ ...props, ...defaultStyledComponents }} />
  );
};

export default ({
  defaultStyleProps,
  createStyledComponents,
  createComponent,
}: $MangoComponent<typeof defaultStyleProps, $StyledProps>);
