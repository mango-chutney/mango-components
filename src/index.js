// @flow

import AnchorButton from './components/AnchorButton';
import Avatar from './components/Avatar';
import Button from './components/Button';
import Checkbox from './components/Checkbox';
import Input from './components/Input';
import MadeWithMangoChutney from './components/MadeWithMangoChutney';
import PaddedCheckbox from './components/PaddedCheckbox';
import { Panel, PanelHeading, PanelSection } from './components/Panel';
import {
  RadialProgressBar,
  NestedRadialProgressBars,
} from './components/RadialProgressBar';
import Tootlip from './components/Tootlip';
import Tristicon from './components/Tristicon';
import Toast from './components/Toast';

export {
  AnchorButton,
  Avatar,
  Button,
  Checkbox,
  Input,
  MadeWithMangoChutney,
  NestedRadialProgressBars,
  PaddedCheckbox,
  Panel,
  PanelHeading,
  PanelSection,
  RadialProgressBar,
  Toast,
  Tootlip,
};

export default {
  AnchorButton,
  Avatar,
  Button,
  Checkbox,
  Input,
  MadeWithMangoChutney,
  NestedRadialProgressBars,
  PaddedCheckbox,
  Panel,
  PanelHeading,
  PanelSection,
  RadialProgressBar,
  Toast,
  Tootlip,
  Tristicon,
};
