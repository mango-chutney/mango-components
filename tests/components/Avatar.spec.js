// @flow

import 'raf/polyfill';
import 'jest-styled-components';
import React from 'react';
import renderer from 'react-test-renderer';
import { createComponent as createAvatarComponent } from '../../src/Avatar';

const Avatar = createAvatarComponent();

function AvatarComponent() {
  return (
    <div>
      <Avatar />
      <Avatar name="test" />
      <Avatar backgroundImage="https://placeholdit.imgix.net/~text?txtsize=33&txt=&w=50&h=50" />
      <Avatar width="200px" />
      <Avatar fontSize="40px" />
    </div>
  );
}

it('Avatar component renders correctly', () => {
  const tree = renderer.create(<AvatarComponent />).toJSON();
  expect(tree).toMatchSnapshot();
});
