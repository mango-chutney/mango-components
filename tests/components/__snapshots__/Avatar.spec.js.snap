// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar component renders correctly 1`] = `
.c0 {
  background-color: #00b4ff;
  background-position: center;
  background-repeat: no-repeat;
  background-size: cover;
  border-radius: 50%;
  height: 2.875rem;
  width: 2.875rem;
  line-height: 2.875rem;
  text-align: center;
  color: #fefefe;
  font-weight: 600;
  text-transform: uppercase;
  font-size: 1.125rem;
}

.c1 {
  background-color: #00b4ff;
  background-position: center;
  background-repeat: no-repeat;
  background-size: cover;
  border-radius: 50%;
  height: 200px;
  width: 200px;
  line-height: 200px;
  text-align: center;
  color: #fefefe;
  font-weight: 600;
  text-transform: uppercase;
  font-size: 1.125rem;
}

.c2 {
  background-color: #00b4ff;
  background-position: center;
  background-repeat: no-repeat;
  background-size: cover;
  border-radius: 50%;
  height: 2.875rem;
  width: 2.875rem;
  line-height: 2.875rem;
  text-align: center;
  color: #fefefe;
  font-weight: 600;
  text-transform: uppercase;
  font-size: 40px;
}

<div>
  <div
    className="c0"
    fontSize="1.125rem"
    style={
      Object {
        "backgroundColor": undefined,
        "backgroundImage": undefined,
      }
    }
    width="2.875rem"
  />
  <div
    className="c0"
    fontSize="1.125rem"
    style={
      Object {
        "backgroundColor": "#8151f3",
        "backgroundImage": undefined,
      }
    }
    width="2.875rem"
  >
    t
  </div>
  <div
    className="c0"
    fontSize="1.125rem"
    style={
      Object {
        "backgroundColor": undefined,
        "backgroundImage": "url(https://placeholdit.imgix.net/~text?txtsize=33&txt=&w=50&h=50)",
      }
    }
    width="2.875rem"
  />
  <div
    className="c1"
    fontSize="1.125rem"
    style={
      Object {
        "backgroundColor": undefined,
        "backgroundImage": undefined,
      }
    }
    width="200px"
  />
  <div
    className="c2"
    fontSize="40px"
    style={
      Object {
        "backgroundColor": undefined,
        "backgroundImage": undefined,
      }
    }
    width="2.875rem"
  />
</div>
`;
